ciao-controller/instance.go

```diff
@@ -191,80 +222,51 @@ func getStorage(c *controller, wl *types.Workload, tenant string, instanceID str
 		device, err := c.CreateBlockDeviceFromSnapshot(s.SourceID, "ciao-image")
 		if err != nil {
 			glog.Errorf("Unable to get block device for image: %v", err)
-			return payloads.StorageResources{}, err
+			return payloads.StorageResource{}, err
 		}
 
-		// don't you need to add support for indicating whether
-		// a block device is bootable.
-		data := types.BlockData{
-			BlockDevice: device,
-			Size:        s.Size,
-			CreateTime:  time.Now(),
-			TenantID:    tenant,
-			Name:        fmt.Sprintf("Storage for instance: %s", instanceID),
-		}
-
-		err = c.ds.AddBlockDevice(data)
-		if err != nil {
-			c.DeleteBlockDevice(device.ID)
-			return payloads.StorageResources{}, err
-		}
+		s.Persistent = true
+		return addBlockDevice(c, tenant, instanceID, device, s)
 
-		return payloads.StorageResources{ID: data.ID, Bootable: s.Bootable, Ephemeral: true}, nil
 	case types.VolumeService:
 		device, err := c.CopyBlockDevice(s.SourceID)
 		if err != nil {
-			return payloads.StorageResources{}, err
+			return payloads.StorageResource{}, err
 		}
 
-		// don't you need to add support for indicating whether
-		// a block device is bootable.
-		data := types.BlockData{
-			BlockDevice: device,
-			Size:        s.Size,
-			CreateTime:  time.Now(),
-			TenantID:    tenant,
-			Name:        fmt.Sprintf("Storage for instance: %s", instanceID),
-		}
-
-		err = c.ds.AddBlockDevice(data)
-		if err != nil {
-			c.DeleteBlockDevice(device.ID)
-			return payloads.StorageResources{}, err
-		}
-
-		return payloads.StorageResources{ID: data.ID, Bootable: s.Bootable}, nil
+		s.Persistent = false
+		return addBlockDevice(c, tenant, instanceID, device, s)
 
 	case types.Empty:
 		device, err := c.CreateBlockDevice("", "", s.Size)
 		if err != nil {
-			return payloads.StorageResources{}, err
+			return payloads.StorageResource{}, err
 		}
 
-		// don't you need to add support for indicating whether
-		// a block device is bootable.
-		data := types.BlockData{
-			BlockDevice: device,
-			Size:        s.Size,
-			CreateTime:  time.Now(),
-			TenantID:    tenant,
-			Name:        fmt.Sprintf("Storage for instance: %s", instanceID),
-		}
+		s.Bootable = false
+		s.Persistent = false
+		return addBlockDevice(c, tenant, instanceID, device, s)
 
-		err = c.ds.AddBlockDevice(data)
-		if err != nil {
-			c.DeleteBlockDevice(device.ID)
-			return payloads.StorageResources{}, err
-		}
+	}
 
-		return payloads.StorageResources{ID: data.ID, Bootable: false}, nil
+	return payloads.StorageResource{}, errors.New("Unsupported workload storage variant in getStorage()")
+}
 
-	}
+func controllerStorageResourceFromPayload(volume payloads.StorageResource) (s types.StorageResource) {
+	s.ID = volume.ID
+	s.Bootable = volume.Bootable
+	s.Persistent = volume.Ephemeral
```

Comment:
Again here..persistence is the apposite ephemerality.

I think we should just standardise on one name.